src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -906,5 +914,141 @@ private static void ReadToEndOfCurrentObject(JsonReader reader)
+                // Your basic project.razor.json using our template have ~800 entries,
+                // so lets start the dictionary out large
+                private int _capacity = 400;
+
+                // This might not acctually need to be concurrent?
+                private readonly ConcurrentDictionary<TKey, WeakReference> _dict = new ConcurrentDictionary<TKey, WeakReference>();
+
+                public bool TryGetValue(TKey key, out TValue value)
+                {
+                    if (key is null)
+                    {
+                        throw new ArgumentNullException(nameof(key));
+                    }
+
+                    if (_dict.TryGetValue(key, out var weak))
+                    {
+                        value = weak.Target as TValue;
+
+                        if (value is null)
+                        {
+                            // This means we lost all references to the value, time to remove the entry.
+                            _dict.TryRemove(key, out _);
+                            return false;
+                        }
+
+                        return true;
+                    }
+
+                    // We didn't find the value
+                    value = null;
+                    return false;
+                }
+
+                public void Add(TKey key, TValue value)
+                {
+                    if (key is null)
+                    {
+                        throw new ArgumentNullException(nameof(key));
+                    }
+
+                    if (value is null)
+                    {
+                        throw new ArgumentNullException(nameof(value));
+                    }
+
+                    if (_dict.Count == _capacity)
+                    {
+                        Cleanup();
+
+                        if (_dict.Count == _capacity)
+                        {
+                            _capacity = _dict.Count * 2;
+                        }
+                    }
+
+                    var weak = new WeakReference(value);
+                    _dict.TryAdd(key, weak);
+                }
+
+                private void Cleanup()
+                {
+                    IList<TKey> remove = null;
+                    foreach (var kvp in _dict)
+                    {
+                        if (kvp.Value is null)
+                        {
+                            // Entry exists but has null value (should never happen)
+                            continue;
+                        }
+
+
+                        if (!(kvp.Value.Target is TValue target))
+                        {
+                            // No references remain for this, remove it from the list entirely
+                            remove ??= new List<TKey>();
+
+                            remove.Add(kvp.Key);
+                        }
+                    }
+
+                    if (remove != null)
+                    {
+                        foreach (var key in remove)
```

Comment:
I'd do a normal for loop since remove is a list. Or you could iterate over the dictionary backwards by going through each of its keys. That way you could remove as you iterate without allocating a list